src/main/java/com/artipie/asto/etcd/EtcdStorage.java

```diff
@@ -0,0 +1,191 @@
+        ).thenApply(rsp -> rsp.getCount() > 0);
+    }
+
+    @Override
+    public CompletableFuture<Collection<Key>> list(final Key prefix) {
+        return this.client.getKVClient().get(
+            keyToSeq(prefix),
+            GetOption.newBuilder()
+                .withKeysOnly(true)
+                .withSortOrder(SortOrder.ASCEND)
+                .build()
+        ).thenApply(
+            rsp -> rsp.getKvs().stream()
+                .map(kv -> new String(kv.getKey().getBytes(), StandardCharsets.UTF_8))
+                .map(str -> new Key.From(str))
+                .distinct()
+                .collect(Collectors.toList())
+        );
+    }
+
+    @Override
+    @SuppressWarnings("PMD.OnlyOneReturn")
+    public CompletableFuture<Void> save(final Key key, final Content content) {
+        final long size = content.size().orElse(0L);
+        if (size <= 0 || size > EtcdStorage.MAX_SIZE) {
+            return new CompletableFutureSupport.Failed<Void>(
+                new IllegalStateException(
+                    String.format("Content size must be in range (1;%d)", EtcdStorage.MAX_SIZE)
+                )
+            ).get();
+        }
+        return new Concatenation(content).single()
+            .map(buf -> new Remaining(buf).bytes())
+            .map(bytes -> ByteSequence.from(bytes))
+            .to(SingleInterop.get())
+            .thenCompose(data -> this.client.getKVClient().put(keyToSeq(key), data))
+            .thenApply(ignore -> (Void) null).toCompletableFuture();
+    }
+
+    @Override
+    public CompletableFuture<Void> move(final Key source, final Key destination) {
+        return this.value(source)
+            .thenCompose(data -> this.save(destination, data))
+            .thenCompose(none -> this.delete(source));
+    }
+
+    @Override
+    public CompletableFuture<Long> size(final Key key) {
+        return this.client.getKVClient().get(keyToSeq(key)).thenApply(
+            rsp -> rsp.getKvs().stream().max(
+                (left, right) -> Long.compare(left.getVersion(), right.getVersion())
+            )
+        ).thenApply(
+            kv -> kv.orElseThrow(
+                () -> new KeyNotFoundException(key)
+            ).getValue().getBytes()
+        ).thenApply(bytes -> Long.valueOf(bytes.length));
+    }
+
+    @Override
+    public CompletableFuture<Content> value(final Key key) {
+        return this.client.getKVClient().get(keyToSeq(key)).thenApply(
+            rsp -> rsp.getKvs().stream().max(
+                (left, right) -> Long.compare(left.getVersion(), right.getVersion())
+            )
+        ).thenApply(
+            kv -> kv.orElseThrow(
+                () -> new KeyNotFoundException(key)
+            ).getValue().getBytes()
+        ).thenApply(bytes -> new Content.From(bytes));
+    }
+
+    @Override
+    public CompletableFuture<Void> delete(final Key key) {
+        return this.client.getKVClient().delete(keyToSeq(key)).thenApply(ignore -> (Void) null);
```

Comment:
@g4s8 hm, will this method fail if key does not exist? It should according to `Storage` contract